controller_manager/scripts/controller_group

```diff
@@ -0,0 +1,132 @@
+# to any other group at run time.
+#
+# Author: Yong Li
+#
+import rospy
+from argparse import ArgumentParser
+from controller_manager import controller_manager_interface
+from controller_manager_msgs.srv import ListControllers, ListControllersRequest
+
+
+def _read_group(group):
+    ''''
+    Return 2 sets of controllers. The first set includes controllers required for the given group,
+    and the second set includes the controllers that should be stopped if the given group is chosen.
+    '''
+    groups = rospy.get_param('controller_groups', {})
+    desired_controllers = set(groups.get(group, ()))
+    if not desired_controllers:
+        rospy.logerr('There is no controller defined in group {}'.format(group))
+        return None, None
+    all_controllers = set()
+    for controllers in groups.values():
+        # Convert list to set to remove duplicates.
+        all_controllers = all_controllers | set(controllers)
+    return desired_controllers, all_controllers - desired_controllers
+
+
+def _get_loaded_controllers():
+    rospy.wait_for_service('controller_manager/list_controllers')
+    s = rospy.ServiceProxy('controller_manager/list_controllers', ListControllers)
+    resp = s.call(ListControllersRequest())
+    controllers = {}
+    for controller in resp.controller:
+        controllers[controller.name] = controller.state == 'running'
+        if controller.state not in ('running', 'stopped'):
+            rospy.logerr('Unknown controller state {}'.format(controller.state))
+    return controllers
+
+
+def _list_group(args):
+    groups = rospy.get_param('controller_groups', {})
+    print('Number of groups: {}'.format(len(groups)))
+    for group, controllers in sorted(groups.items()):
+        print('  {}'.format(group))
+        for controller in controllers:
+            print('      {}'.format(controller))
+
+
+def _spawn_group(args):
+    desired_controllers, _ = _read_group(args.group)
+    if not desired_controllers:
+        return
+    for controller in desired_controllers:
+        controller_manager_interface.load_controller(controller)
+    for controller in desired_controllers:
+        controller_manager_interface.start_controller(controller)
+
+
+def _switch_group(args):
+    desired_controllers, other_controllers = _read_group(args.group)
+    if not desired_controllers:
+        return
+    loaded_controllers = _get_loaded_controllers()
+
+    # Stop running controllers that we no longer need.
+    stop = []
+    for controller in other_controllers:
+        if loaded_controllers.get(controller, False):
+            stop.append(controller)
+
+    # Load and start every desired controller if needed.
+    start = []
+    for controller in desired_controllers:
+        if controller not in loaded_controllers:
+            controller_manager_interface.load_controller(controller)
+        if not loaded_controllers.get(controller, False):
+            start.append(controller)
```

Comment:
Why is there a conditional here? If a controller that's needed couldn't be loaded, the whole operation should fail, right?

Reply:
Oh, got it— so you're figuring out the stops/starts required to get from the current state to the desired state, as opposed to just stop everything and start <desired>.